src/rtc.rs

```diff
@@ -1,210 +1,343 @@
-                    .clear_bit()
-            });
-            
+            init_mode(&self.rtc, false);
         }
-        write_protection(&rtc, true);
+        write_protection(&self.rtc, true);
+    }
 
-        Self {
-            rtc: rtc
+    /// Set Time
+    /// Note: If setting both time and date, use set_date_time(...) to avoid errors.
+    pub fn set_time(&self, time: Time) {
+        write_protection(&self.rtc, false);
+        {
+            init_mode(&self.rtc, true);
+            {
+                set_time_raw(&self.rtc, time);
+            }
+            init_mode(&self.rtc, false);
         }
+        write_protection(&self.rtc, true);
     }
 
-    pub fn set_time(&self, time: &Time){
+    /// Set Date
+    /// Note: If setting both time and date, use set_date_time(...) to avoid errors.
+    pub fn set_date(&self, date: Date) {
         write_protection(&self.rtc, false);
         {
             init_mode(&self.rtc, true);
             {
-                
-                let (ht, hu) = byte_to_bcd2(time.hours as u8);
-                let (mnt, mnu) = byte_to_bcd2(time.minutes as u8);
-                let (st, su) = byte_to_bcd2(time.seconds as u8);
-                self.rtc.tr.write(|w| unsafe {
-                    w.ht().bits(ht)
-                        .hu().bits(hu)
-                        .mnt().bits(mnt)
-                        .mnu().bits(mnu)
-                        .st().bits(st)
-                        .su().bits(su)
-                        .pm()
-                        .clear_bit()
-
-                });
-
-                self.rtc.cr.modify(|_, w| {
-                    w.fmt()
-                        .bit(time.daylight_savings)
-
-                });
+                set_date_raw(&self.rtc, date);
             }
             init_mode(&self.rtc, false);
         }
         write_protection(&self.rtc, true);
     }
 
-    pub fn get_time(&self) -> Time {
-        let time;
-        
-        let timer = self.rtc.tr.read();
-        let cr = self.rtc.cr.read();
-        time = Time::new(bcd2_to_byte((timer.ht().bits(), timer.hu().bits())).into(), 
-                        bcd2_to_byte((timer.mnt().bits(), timer.mnu().bits())).into(),
-                        bcd2_to_byte((timer.st().bits(), timer.su().bits())).into(),
-                        cr.fmt().bit());
-        
-        write_protection(&self.rtc, true);
-        
-        time
+    pub fn get_config(&self) -> RtcConfig {
+        self.rtc_config
     }
 
-    pub fn set_date(&self, date: &Date){
+    /// Applies the RTC config
+    /// It this changes the RTC clock source the time will be reset
+    pub fn set_config(
+        &mut self,
+        bdcr: &mut BDCR,
+        pwrcr1: &mut pwr::CR1,
+        rtc_config: RtcConfig,
+    ) {
+        // Unlock the backup domain
+        pwrcr1.reg().modify(|_, w| w.dbp().set_bit());
+        while pwrcr1.reg().read().dbp().bit_is_clear() {}
+
+        let reg = bdcr.enr().read();
+        assert!(
```

Comment:
This is dependent on a register, will never compile away. Should probably be changed

Reply:
I'm not sure what you mean by never compile away, or how to fix it..

Reply:
It should be done in a better way but i have yet to find any. But it should also be possible to make the RTC compatible with CSS. I have yet to find a solution to either.